### ISSUE TITLE
Doesn't neuron potential reset to 0 once spike occurs?

### ISSUE THREAD

**Issue by sms821** [NeuromorphicProcessorProject/snn_toolbox#31]
I ran the simulation for the lenet5 example using poisson inputs on a single image. The poisson spike train was of duration 10 timesteps. I observed the values of 'mem_n_b_l_t' of the last layer and saw that the membrane potential does not reset to 0 after a spike occurs. In fact, it remains the same until, in the next time step, a weighted sum of the input spikes is added to it. 

So the membrane potential either remains constant(no spike) or increases over its earlier value(spike). This is in contrast with my understanding that membrane potential should reset to 0 once the threshold is reached. Also, isn't the threshold supposed to be fixed at 1 for all timesteps?

My questions might be naive but I am trying hard to understand how SNNs work. I would highly appreciate it if you could clarify this part for me.
Thanks!

**Comment by rbodo** [comment]
> This is in contrast with my understanding that membrane potential should reset to 0 once the threshold is reached. Also, isn't the threshold supposed to be fixed at 1 for all timesteps?

This is correct. (There is an option in INIsim to reset not to zero but to the excess charge above threshold, called "reset by subtraction". There the membrane potential is reduced by the threshold value at the time of a spike.)

I have noticed weird artifacts in the membrane potential plots before. I attribute them to matplotlib not being able to show the overlapping traces of a large number of neurons, so that sometimes the reset lines and others disappear. But I can't say if it's the case here without seeing the plot.

Try to compare the potential plot against the spiketrain plot. Do the times when the membrane potential crosses threshold have a corresponding dot in the rasterplot? 

By the way, which simulator are you using, the builtin INIsim (default) or pyNN? Is the accuracy of the SNN ok?

**Comment by sms821** [comment]
Hello, Thanks for your reply. 

I am using the default INIsim simulator and SNN accuracy 100% for the single input image. I ran the same simulation again, but this time, changed the reset config to "reset to zero". I also tried plotting the neuron membrane potential by setting "plot_vars = {'v_mem'}" in the config file. However, the tool failed to plot any membrane potential data (it's plotting the other metrics such as correlation, spikerate etc. ok). As a result, I am unable to compare it against the spike train raster plot.

**Comment by rbodo** [comment]
Oh, I see now - you mentioned that you were looking at the output layer. This layer is special in that it has a softmax nonlinearity. We have different ways of implementing this nonlinearity in INIsim; the default one uses stochastic neurons that just fire in proportion to their membrane potential, without resetting. In the other layers, you should observe the reset as expected.

And yes, when using INIsim, plotting the v_mem is not supported. You can plot the membrane potential when using pyNN simulators.

**Comment by sms821** [comment]
Thanks a lot for the clarification! 

It saved me a lot of time as I am trying to port one of these applications to a hardware simulator for spiking neural networks to study their hardware-related properties. 

This is a great tool for people trying to understand and use SNNs .
